scipy/stats/tests/test_sobol_.py

```diff
@@ -0,0 +1,219 @@
+            self.draws_scrambled_1d = engine_scrambled_1d.random(10)
+            engine_scrambled_3d = Sobol(3, scramble=True, seed=12345)
+            self.draws_scrambled_3d = engine_scrambled_3d.random(10)
+
+    def test_warning(self):
+        with pytest.warns(UserWarning):
+            engine = Sobol(1)
+            engine.random(10)
+
+    def test_random_base2(self):
+        self.setUp()
+        engine = Sobol(1, scramble=False)
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[:4],
+                           sample)
+
+        # resampling still having N=2**n
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[4:8],
+                           sample)
+
+        # resampling again but leading to N!=2**n
+        assert_raises(ValueError, engine.random_base2, 2)
+
+    def test_raise(self):
+        assert_raises(ValueError, Sobol, Sobol.MAXDIM + 1)
+
+    def test_Unscrambled1DSobol(self):
+        self.setUp()
+        expected = [
+            0, 0.5, 0.75, 0.25, 0.375, 0.875, 0.625, 0.125, 0.1875, 0.6875
+        ]
+        assert_equal(self.draws_unscrambled_1d.shape[0], 10)
+        assert_equal(self.draws_unscrambled_1d.shape[1], 1)
+        assert_array_equal(self.draws_unscrambled_1d.flatten(),
+                           np.array(expected))
+
+    def test_Unscrambled3DSobol(self):
+        self.setUp()
+        expected_dim3 = [0, 0.5, 0.25, 0.75, 0.625, 0.125, 0.875, 0.375,
+                         0.9375, 0.4375]
+        assert_equal(self.draws_unscrambled_3d.shape[0], 10)
+        assert_equal(self.draws_unscrambled_3d.shape[1], 3)
+        assert_array_equal(self.draws_unscrambled_3d[:, 2],
+                           np.array(expected_dim3))
+        assert_array_equal(
+            self.draws_unscrambled_3d[:, 0],
+            self.draws_unscrambled_1d.flatten()
+        )
+
+    def test_Unscrambled3DAsyncSobol(self):
+        self.setUp()
+        engine_unscrambled_3d = Sobol(3, scramble=False)
+        draws = np.vstack([engine_unscrambled_3d.random() for i in range(10)])
+        assert_array_equal(self.draws_unscrambled_3d, draws)
+
+    def test_UnscrambledFastForwardAndResetSobol(self):
+        self.setUp()
+        engine_unscrambled_3d = Sobol(3, scramble=False).fast_forward(5)
+        draws = engine_unscrambled_3d.random(5)
+        assert_array_equal(self.draws_unscrambled_3d[5:10, :], draws)
+
+        engine_unscrambled_3d.reset()
+        even_draws = []
+        for i in range(10):
+            if i % 2 == 0:
+                even_draws.append(engine_unscrambled_3d.random())
+            else:
+                engine_unscrambled_3d.fast_forward(1)
+        assert_array_equal(
+            self.draws_unscrambled_3d[[i for i in range(10) if i % 2 == 0]],
+            np.vstack(even_draws),
+        )
+
+    def test_UnscrambledHighDimSobol(self):
```

Comment:
Similarly here - is there a reference I can use to check that this is correct?

Reply:
This I am actually not sure. @Balandat you know this?

Reply:
We're essentially checking that the first draw in this 1111-dim space is the origin, and the second draw is `[0.5]^1111`. The third assertion ensures that the next points in the sequence are filling in the dimensions with 0.25 and 0.75 though I'm not sure about the numbers 557 and 554. 

I think that @ArtOwen or @fjhickernell would have the best understanding of what a good reference for this would be.

Reply:
I'm resolving this now because all the tests have been combined into one file and I can't easily find this test anymore. When I review Sobol specifically, we'll see if there are any tests without references.